### ISSUE TITLE
how to do the expansion?

### ISSUE THREAD

**Issue by Xiaocong6** [google-research/morph-net#75]
hi, I have a question about the expansion.
I trained the regularizer_loss and got the following network structure.
How should I use a uniform width multiplier to do the expansion?
![微信截图_20190524143237](https://user-images.githubusercontent.com/44869509/58307316-da72d700-7e30-11e9-9105-2fc074a25cd1.png)
Do I need to manually adjust the number of channels, if I use the multiplier of 2, does that mean I need to set the number of  the layer('g_net/enc1_2/conv1/Conv2D: 12')'s channels to 24.

Also, I would like to ask if morph-net can be used for pixel to pixel tasks, such as super resolution and de-blur. Can keep the same accuracy after shrinking the network?

**Comment by ayp-google** [comment]
Right, you can expand by multiplying the number in the network structure (e.g., 12 -> 24 in your example).  It is usually easier to write a short script to do this.  Note that after expansion, it is possible to have inconsistent sizes due to rounding if you use a non-integer factor such as 1.2.  If so, the graph will fail to build and you will need to manually change the value to correct for the rounding.  Hope that helps.

We haven't tried using MorphNet in super resolution or deblur tasks yet.  If the network architecture is a conv net, then MorphNet should still work the same.  The accuracy loss will depend on the network architecture and how aggressively you shrink.  If you try this, we would be happy to discuss the result with you.

**Comment by yairmov** [comment]
To add on the comment above - While we haven't applied MorphNet to super resolution or deblur tasks, we have used it on semantic segmentation U-Net like models which are similar to super resolution, and had good results. As mentioned above, the accuracy loss will depend on your specific task/data and the amount of shrinking.

**Comment by Xiaocong6** [comment]
Thanks! It is so clear.
I used the Morph-net to shrink the SRN-deblur network. 
The loss: 
network_regularizer = flop_regularizer.GroupLassoFlopsRegularizer([x_unwrap.op], threshold=1e-3)
regularization_strength = 1e-10
regularizer_loss = (network_regularizer.get_regularization_term() * regularization_strength)

The checkpoint file of the original model is about 80MB. I chose the structure generated by step12000, and the checkpoint of the small model is only about 20MB. I compared the deblurring results of these two models. The result of the small network are obviously poor. I am going to try the structure of 10,000 step or 5000 step or 8000 step and so on, or to extend the generated structure. Do you think this is okay, Any other opinions based on your experience?

**Comment by ayp-google** [comment]
One set of experiments is to use the structure at different steps as you are planning to do.  Another option is to adjust the regularization_strength.  If the shrunk model is too small, then you can try a weaker regularization_strength such as 5e-11.  You may need to try more values to get the right accuracy/cost tradeoff.

Also, if your target is model size, you can use model_size_regularizer.GroupLassoModelSizeRegularizer to get a better accuracy/cost tradeoff.  You can read the paper for more details, but the summary is that targeting the wrong resource can result in a bad tradeoff for the cost you care about (e.g., model size vs. FLOPs).

**Comment by H-arshit** [comment]
Hello @yairmov, 
As mentioned in your comment above, you guys have achieved good results on UNet,
i was working on a similar problem for UNet with ModelSizeRegularizer, and am facing some problem with that.
Could you please open source the implementation of Morphnet over UNet.

**Comment by yairmov** [comment]
There shouldn't be anything special needed to run Morphnet on a UNet-type model. Provide the outputs of your model to the model size regularizer and our framework will assign size costs to all the convolution ops.
